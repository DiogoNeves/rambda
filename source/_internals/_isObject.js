import { _isArray } from "./_isArray";

export function _isObject (item){
  if (!item) return false;

  return typeof item === 'object' && !_isArray(item);
}
