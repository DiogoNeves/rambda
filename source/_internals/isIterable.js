import {type} from '../type'

export function _isIterable(input) {
  return ['Object', 'Array'].includes(type(input))
}
