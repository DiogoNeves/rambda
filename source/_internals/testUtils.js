import combinate from 'combinate'
import { equals, filter, map, omit, switcher, type } from 'rambdax'

const omitOk = omit('ok')

const PENDING = 'PENDING'
const RESULTS_EQUAL = 'results are equal'
const RESULTS_MISMATCH = 'results are different'
const ERRORS_EQUAL = 'errors are equal'
const ERRORS_MISMATCH = 'errors are different'
const SHOULD_THROW = 'Rambda should throw'
const SHOULD_NOT_THROW = 'Rambda should not throw'

const MISSING = 'Missing error handle in parseError'

function parseError(err){
  const typeError = switcher(err)
    .is(x => x instanceof TypeError, 'TypeError')
    .is(x => x instanceof TypeError, 'TypeError')
    .is(x => x instanceof TypeError, 'TypeError')
    .default(MISSING)

  if (typeError === MISSING){
    throw new Error('typeError === MISSING')
  }

  return {
    message : err.message,
    type    : typeError,
    ok      : true,
  }
}

export function show(input){
  const typeInput = type(input)
  if ([ 'Promise', 'Async' ].includes(typeInput)){
    return ''
  }

  if (typeInput === 'Array'){
    if (input.length === 0) return '[]'

    return `[${ input.map(show).join(', ') }]`
  }

  if (typeInput === 'Object'){
    if (Object.keys(input).length === 0) return '{}'

    return JSON.stringify(map(show, input))
  }
  if ([ 'Boolean', 'Number', 'String' ].includes(typeInput)){
    return input
  }
  if ([ 'Null', 'Undefined' ].includes(typeInput)){
    return typeInput.toLowerCase()
  }

  return input.toString()
}

function executeSync(fn, inputs){
  let result = PENDING
  let error = { ok : false }
  try {
    result = fn(...inputs)
  } catch (e){
    error = parseError(e)
  }

  return {
    result,
    error,
  }
}

async function executeAsync(fn, inputs){
  let result = PENDING
  let error = { ok : false }
  try {
    result = await fn(...inputs)
  } catch (e){
    error = parseError(e)
  }

  return {
    result,
    error,
  }
}

export function profileMethod(
  firstInput,
  secondInput = undefined,
  thirdInput = undefined,
  fn
){
  const combinationsInput = filter(Boolean, {
    firstInput,
    secondInput,
    thirdInput,
  })
  const inputKeys = Object.keys(combinationsInput)
  const combinations = combinate(combinationsInput)

  combinations.forEach(combination => {
    const inputs = [
      combination.firstInput,
      combination.secondInput,
      combination.thirdInput,
    ].filter((_, i) => i < inputKeys.length)

    test(getTestTitle(...inputs), () => {
      const { result, error } = executeSync(fn, inputs)

      expect({
        result,
        error : error.ok ? omitOk(error) : PENDING,
        inputs,
      }).toMatchSnapshot()
    })
  })
}

export function profileMethodAsync({
  firstInput,
  secondInput = undefined,
  thirdInput = undefined,
  fn,
  callback,
}){
  afterAll(() => callback())
  const combinationsInput = filter(Boolean, {
    firstInput,
    secondInput,
    thirdInput,
  })
  const inputKeys = Object.keys(combinationsInput)
  const combinations = combinate(combinationsInput)

  combinations.forEach(combination => {
    const inputs = [
      combination.firstInput,
      combination.secondInput,
      combination.thirdInput,
    ].filter((_, i) => i < inputKeys.length)
    test(getTestTitle(...inputs), async () => {
      const { result, error } = await executeAsync(fn, inputs)

      expect({
        result,
        error : error.ok ? omitOk(error) : PENDING,
        inputs,
      }).toMatchSnapshot()
    })
  })
}

export function compareToRamda(fn, fnRamda){
  return (...inputs) => {
    const { result, error } = executeSync(fn, inputs)
    const { result: ramdaResult, error: ramdaError } = executeSync(fnRamda,
      inputs)

    const toReturn = {
      result,
      ramdaResult,
      ramdaError : ramdaError.ok ? omitOk(ramdaError) : PENDING,
      error      : error.ok ? omitOk(error) : PENDING,
    }

    if (result !== PENDING){
      if (ramdaError.ok){
        return {
          ...toReturn,
          ok    : false,
          label : SHOULD_THROW,
        }
      }

      if (equals(result, ramdaResult)){
        return {
          ...toReturn,
          ok    : true,
          label : RESULTS_EQUAL,
        }
      }

      return {
        ...toReturn,
        ok    : false,
        label : RESULTS_MISMATCH,
      }
    }

    if (equals(error, ramdaError)){
      return {
        ...toReturn,
        ok    : true,
        label : ERRORS_EQUAL,
      }
    }

    if (ramdaError.ok){
      return {
        ...toReturn,
        ok    : false,
        label : ERRORS_MISMATCH,
      }
    }

    return {
      ...toReturn,
      ok    : false,
      label : SHOULD_NOT_THROW,
    }
  }
}

export const getTestTitle = (...inputs) =>
  inputs.map(x => `${ type(x) } ${ show(x) }`).join(' | ')

export const compareCombinations = ({
  firstInput,
  secondInput = undefined,
  thirdInput = undefined,
  setCounter = () => {},
  setGlobalCounter = () => {},
  fn,
  fnRamda,
}) => {
  const combinationsInput = filter(Boolean, {
    firstInput,
    secondInput,
    thirdInput,
  })
  const inputKeys = Object.keys(combinationsInput)
  const combinations = combinate(combinationsInput)
  const compareOutputs = compareToRamda(fn, fnRamda)

  combinations.forEach(combination => {
    const inputs = [
      combination.firstInput,
      combination.secondInput,
      combination.thirdInput,
    ].filter((_, i) => i < inputKeys.length)

    test(getTestTitle(...inputs), () => {
      const compared = compareOutputs(...inputs)
      setGlobalCounter()
      if (!compared.ok){
        setCounter()
        expect({
          ...compared,
          inputs,
        }).toMatchSnapshot()
      }
    })
  })
}
