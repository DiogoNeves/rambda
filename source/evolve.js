import { mapArray, mapObject } from './map'
import { type } from './type'

export function evolveArray(rules, list){
  if (!list){
    return {}
  }

  return mapArray(
    (x, i) => {
      if (type(rules[ i ]) === 'Function'){
        return rules[ i ](x)
      }

      return x
    },
    list,
    true
  )
}

export function evolveObject(rules, input){
  const inputType = type(input)
  if (inputType !== 'Object') return {}

  return mapObject((x, prop) => {
    if (type(x) === 'Object'){
      const typeRule = type(rules[ prop ])
      if (typeRule === 'Function'){
        return rules[ prop ](x)
      }
      if (typeRule === 'Object'){
        return evolve(rules[ prop ], x)
      }

      return x
    }
    if (type(rules[ prop ]) === 'Function'){
      return rules[ prop ](x)
    }

    return x
  }, input)
}

export function evolve(rules, input){
  if (arguments.length === 1){
    return _input => evolve(rules, _input)
  }
  const rulesType = type(rules)
  if (rulesType === 'Array') return evolveArray(rules, input)
  if (rulesType === 'Object') return evolveObject(rules, input)
  const inputType = type(input)

  if (inputType === 'Array'){
    if (input.length > 0){
      throw new TypeError(`Cannot read property '0' of ${ rules }`)
    }

    return []
  }

  return {}
}
