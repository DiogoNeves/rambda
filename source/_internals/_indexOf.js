import { equals } from '../equals'
import {_isObject} from './_isObject'

/**
 * Get first index of item in a list based on native `indexOf` or on `R.equals` if input is list of objects.
 */
export function _indexOf(valueToFind, list) {
  if (!_isObject(valueToFind)) return list.indexOf(valueToFind)

  let index = -1
  list.forEach((x, i) => {
    if (index > -1) return
    if (equals(x, valueToFind)) index = i
  })

  return index
}
