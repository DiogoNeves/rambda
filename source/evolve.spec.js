import { evolve as evolveRamda } from 'ramda'

import { add } from '../rambda.js'
import { compareCombinations, compareToRamda } from './_internals/testUtils'
import { evolve } from './evolve'

test('happy', () => {
  const rules = {
    foo : add(1),
    bar : add(-1),
  }
  const input = {
    a   : 1,
    foo : 2,
    bar : 3,
  }
  const result = evolve(rules, input)
  expect(result).toEqual({
    a   : 1,
    foo : 3,
    bar : 2,
  })
})

test('is recursive', () => {
  const transf = {
    nested : {
      second : add(-1),
      third  : add(1),
    },
  }
  const object = {
    first  : 1,
    nested : {
      second : 2,
      third  : 3,
    },
  }
  const expected = {
    first  : 1,
    nested : {
      second : 1,
      third  : 4,
    },
  }
  const result = evolve(transf, object)
  expect(result).toEqual(expected)
})

test('ignores primitive value transformations', () => {
  const transf = {
    n : 2,
    m : 'foo',
  }
  const object = {
    n : 0,
    m : 1,
  }
  const expected = {
    n : 0,
    m : 1,
  }
  const result = evolve(transf, object)
  expect(result).toEqual(expected)
})

test('with array', () => {
  const transf = [ add(1), add(-1) ]
  const list = [ 100, 1400 ]
  const expected = [ 101, 1399 ]
  const result = evolve(transf, list)
  expect(result).toEqual(expected)
})

const rulesObject = { a : add(1) }
const rulesList = [ add(1) ]
const possibleIterables = [ null, undefined, '', 42, [], [ 1 ], { a : 1 } ]
const possibleRules = [ ...possibleIterables, rulesList, rulesObject ]

describe('r.evolve', () => {
  let counter = 1
  let globalCounter = 1
  afterAll(() => {
    console.log({ counter })
    console.log({ globalCounter })
  })
  compareCombinations({
    firstInput       : possibleRules,
    setCounter       : () => counter++,
    setGlobalCounter : () => globalCounter++,
    secondInput      : possibleIterables,
    fn               : evolve,
    fnRamda          : evolveRamda,
  })
})

test.skip('foo', () => {
  const compareOutputs = compareToRamda(evolve, evolveRamda)

  const rulesInput = null
  const iterableInput = [ 1 ]
  // const iterableInput = { a : 1 }
  const compared = compareOutputs(rulesInput, iterableInput)
  console.log(compared)
  console.log(compared.ramdaResult)
  console.log(compared.label)
})
