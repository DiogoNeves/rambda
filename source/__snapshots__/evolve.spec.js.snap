// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`r.evolve Null null | Array  1`] = `
Object {
  "error": "PENDING",
  "inputs": Array [
    null,
    Array [],
  ],
  "label": "results are different",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": Array [],
  "result": Object {},
}
`;

exports[`r.evolve Null null | Array 1 1`] = `
Object {
  "error": "PENDING",
  "inputs": Array [
    null,
    Array [
      1,
    ],
  ],
  "label": "Rambda should throw",
  "ok": false,
  "ramdaError": "Cannot read property '0' of null",
  "ramdaResult": "PENDING",
  "result": Object {},
}
`;

exports[`r.evolve Number 42 | Array  1`] = `
Object {
  "error": "PENDING",
  "inputs": Array [
    42,
    Array [],
  ],
  "label": "results are different",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": Array [],
  "result": Object {},
}
`;

exports[`r.evolve Number 42 | Array 1 1`] = `
Object {
  "error": "PENDING",
  "inputs": Array [
    42,
    Array [
      1,
    ],
  ],
  "label": "results are different",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": Array [
    1,
  ],
  "result": Object {},
}
`;

exports[`r.evolve Object [object Object] | Array  1`] = `
Object {
  "error": "Cannot read property '0' of ",
  "inputs": Array [
    Object {
      "a": 1,
    },
    Array [],
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": Array [],
  "result": "PENDING",
}
`;

exports[`r.evolve Object [object Object] | Array  2`] = `
Object {
  "error": "Cannot read property '0' of ",
  "inputs": Array [
    Object {
      "a": [Function],
    },
    Array [],
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": Array [],
  "result": "PENDING",
}
`;

exports[`r.evolve Object [object Object] | Array 1 1`] = `
Object {
  "error": "Cannot read property '0' of 1",
  "inputs": Array [
    Object {
      "a": 1,
    },
    Array [
      1,
    ],
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": Array [
    1,
  ],
  "result": "PENDING",
}
`;

exports[`r.evolve Object [object Object] | Array 1 2`] = `
Object {
  "error": "Cannot read property '0' of 1",
  "inputs": Array [
    Object {
      "a": [Function],
    },
    Array [
      1,
    ],
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": Array [
    1,
  ],
  "result": "PENDING",
}
`;

exports[`r.evolve String  | Array  1`] = `
Object {
  "error": "Cannot read property '0' of ",
  "inputs": Array [
    "",
    Array [],
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": Array [],
  "result": "PENDING",
}
`;

exports[`r.evolve String  | Array 1 1`] = `
Object {
  "error": "Cannot read property '0' of 1",
  "inputs": Array [
    "",
    Array [
      1,
    ],
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": Array [
    1,
  ],
  "result": "PENDING",
}
`;

exports[`r.evolve Undefined undefined | Array  1`] = `
Object {
  "error": "PENDING",
  "inputs": Array [
    undefined,
    Array [],
  ],
  "label": "results are different",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": Array [],
  "result": Object {},
}
`;

exports[`r.evolve Undefined undefined | Array 1 1`] = `
Object {
  "error": "PENDING",
  "inputs": Array [
    undefined,
    Array [
      1,
    ],
  ],
  "label": "Rambda should throw",
  "ok": false,
  "ramdaError": "Cannot read property '0' of undefined",
  "ramdaResult": "PENDING",
  "result": Object {},
}
`;
